rust/lance/src/io/writer/statistics.rs

```diff
@@ -0,0 +1,1082 @@
+            }
+        }
+        DataType::Float64 => {
+            if scalar_min_value == ScalarValue::Float64(Some(0.0)) {
+                scalar_min_value = ScalarValue::Float64(Some(-0.0));
+            }
+            if scalar_max_value == ScalarValue::Float64(Some(-0.0)) {
+                scalar_max_value = ScalarValue::Float64(Some(0.0));
+            }
+        }
+        DataType::Date32 => {
+            let min_value_date32 = scalar_min_value
+                .to_array()
+                .as_any()
+                .downcast_ref::<Int32Array>()
+                .unwrap()
+                .value(0);
+            let max_value_date32 = scalar_max_value
+                .to_array()
+                .as_any()
+                .downcast_ref::<Int32Array>()
+                .unwrap()
+                .value(0);
+            scalar_min_value = ScalarValue::Date32(Some(min_value_date32));
+            scalar_max_value = ScalarValue::Date32(Some(max_value_date32));
+        }
+        DataType::Timestamp(_, _) => {
+            // TODO: Grab the correct timezone from the array or should we ignore it for stats?
+            // let arr = arrays[0].as_any().downcast_ref::<TimestampMicrosecondArray>().unwrap();
+            // let tz = Some(Arc::new(arr.timezone().unwrap()));
+            let tz = Some("UTC".into());
+
+            let min_value_timestamp = scalar_min_value
+                .to_array()
+                .as_any()
+                .downcast_ref::<Int64Array>()
+                .unwrap()
+                .value(0);
+            let max_value_timestamp = scalar_max_value
+                .to_array()
+                .as_any()
+                .downcast_ref::<Int64Array>()
+                .unwrap()
+                .value(0);
+            scalar_min_value =
+                ScalarValue::TimestampMicrosecond(Some(min_value_timestamp), tz.clone());
+            scalar_max_value =
+                ScalarValue::TimestampMicrosecond(Some(max_value_timestamp), tz.clone());
+        }
+        DataType::Duration(_) => {
+            let min_value_duration = scalar_min_value
+                .to_array()
+                .as_any()
+                .downcast_ref::<Int64Array>()
+                .unwrap()
+                .value(0);
+            let max_value_duration = scalar_max_value
+                .to_array()
+                .as_any()
+                .downcast_ref::<Int64Array>()
+                .unwrap()
+                .value(0);
+            scalar_min_value = ScalarValue::DurationMillisecond(Some(min_value_duration));
+            scalar_max_value = ScalarValue::DurationMillisecond(Some(max_value_duration));
+        }
+        _ => {}
+    }
+
+    StatisticsRow {
+        null_count: ScalarValue::Int64(Some(null_count)),
+        min_value: scalar_min_value,
+        max_value: scalar_max_value,
+    }
+}
+
+fn get_decimal_statistics(arrays: &[&ArrayRef]) -> StatisticsRow {
```

Comment:
I couldn't quite make this work with generic `get_statistics` because of `T::Native::max_value()` not matching `i128::MAX` (although it should). Open to suggestions.

Reply:
Could you explain this more? What is the value of `T::Native::max_value()` and where is that being defined?

Reply:
https://github.com/lancedb/lance/blob/d55559c652c815e0020e9f701dacd3a8a4d2c265/rust/lance/src/io/writer/statistics.rs#L63-L73
The generic `get_statistics<<T: ArrowNumericType>>` function uses `T::Native::max_value();` to get the maximum value of the type. This is how I get from arrow array types to maximum and minimum values in native types. However when doing this for decimal128 rust throws templateing errors that have to do with casting between arrow and native i128 and I decided to not dig too deep yet.